### PR TITLE
[WIP] Cleanup of Wavelet and TSLess stuff

### PR DESCRIPTION
**Pull request by ilfreddy**
## Description
Removal of code related to Wavelets and TsLess cavity which is not present in the released version 

## How Has This Been Tested?
All tests, except the standalone ones now pass

## Todos
Standalone tests shall be fixed

### CODE REVIEW

**Comment by bast** [comment]
The todo part is part of this PR - in other words work in progress or is it to be done after this PR is accepted?

**Comment by ilfreddy** [comment]
The todos is something I have not managed to fix yet. Roberto will have a look. This PR is still WIP, although I have not marked it as such.

**Comment by robertodr** [comment]
Thanks for the PR. In Markdown the  `- [ ]` are rendered as checkboxes, so you can leave them unchecked to signal discussion points and/or WIP status. For example:
- [ ] Not done (rendered from `- [ ] Not done`
- [x] Done (rendered from `- [x] Done`)

 I would prefer that the title of the PR clearly states if it's WIP or not.

File: src/cavity/CavityData.hpp [hunk 1]
```diff
@@ -45,20 +45,11 @@ struct CavityData {
   /*! The minimal distance between two sampling points on different spheres. Relevant
    * for TsLessCavity */
   double minDistance;
```
Comment (robertodr) [inline src/cavity/CavityData.hpp]: This parameter should also be purged!

File: src/cavity/CavityData.hpp [hunk 2]
```diff
@@ -67,22 +58,14 @@ struct CavityData {
...
              double _minRadius,
-             int _patchLevel,
-             double _coarsity,
-             const std::string & _fname,
-             const std::string & _dyad)
+             const std::string & _fname)
       : molecule(_molec),
         area(_area),
         probeRadius(_probeRadius),
         minDistance(_minDistance),
```
Comment (robertodr) [inline src/cavity/CavityData.hpp]: This should be removed too.

File: src/interface/Input.cpp [hunk 3]
```diff
@@ -215,10 +209,7 @@ void Input::reader(const PCMInput & host_input) {
 
   type_ = detail::trim_and_upper(host_input.cavity_type);
   area_ = host_input.area * angstrom2ToBohr2();
-  patchLevel_ = host_input.patch_level;
-  coarsity_ = host_input.coarsity * angstromToBohr();
   minDistance_ = host_input.min_distance * angstromToBohr();
```
Comment (robertodr) [inline src/interface/Input.cpp]: Also this one should be removed.

File: src/cavity/CavityData.hpp [hunk 4]
```diff
@@ -67,22 +58,14 @@ struct CavityData {
              double _area,
              double _probeRadius,
              double _minDistance,
```
Comment (robertodr) [inline src/cavity/CavityData.hpp]: This parameter should be removed from the constructor.

File: src/interface/Input.cpp [hunk 5]
```diff
@@ -366,12 +355,8 @@ CavityData Input::cavityParams() const {
                     area_,
                     probeRadius_,
                     minDistance_,
```
Comment (robertodr) [inline src/interface/Input.cpp]: This parameter should be removed.

File: src/interface/Input.hpp [hunk 6]
```diff
@@ -88,6 +88,7 @@ class Input {
   /// Cavity section input
   std::string cavityType() const { return type_; }
   bool scaling() const { return scaling_; }
+  double cavityArea() const { return area_; }
```
Comment (robertodr) [inline src/interface/Input.hpp]: Why did you add this getter?

File: src/interface/Input.hpp [hunk 7]
```diff
@@ -175,18 +175,10 @@ class Input {
...
   std::string cavFilename_;
-  /// Filename for the wavelet cavity dyadic file
-  std::string dyadicFilename_;
-  /// Wavelet cavity patch level
-  int patchLevel_;
-  /// Wavelet cavity coarsity
-  double coarsity_;
   /// GePol and TsLess cavities average element area
   double area_;
   /// TsLess cavity minimal distance between sampling points
```
Comment (robertodr) [inline src/interface/Input.hpp]: These two lines should be removed as well.

File: tests/input/input_cpcm.cpp [hunk 8]
```diff
@@ -48,14 +48,12 @@ TEST_CASE("Input reading using GetKw for an input file for a TsLess cavity",
...
-  std::string type = "TSLESS";
+  std::string type = "GEPOL";
   double area = 0.6 * angstrom2ToBohr2();
-  double minDistance = 5.0 * angstromToBohr();
-  int derOrder = 25;
   bool scaling = false;
   double diagonalScaling = 1.0694;
   std::string radiiSet = "UFF";
-  double minimalRadius = 0.1 * angstromToBohr();
+  double minimalRadius = 0.3 * angstromToBohr();
```
Comment (robertodr) [inline tests/input/input_cpcm.cpp]: There should be a check added on the minimal radius parameter.

File: tests/input/input_diffuse.cpp [hunk 9]
```diff
@@ -90,8 +88,7 @@ TEST_CASE("Input reading using GetKw for an input file for a diffuse environment
   REQUIRE(units == parsedInput.units());
   REQUIRE(CODATAyear == parsedInput.CODATAyear());
   REQUIRE(type == parsedInput.cavityType());
-  REQUIRE(patchLevel == parsedInput.cavityParams().patchLevel);
-  REQUIRE(coarsity == Approx(parsedInput.cavityParams().coarsity));
+  REQUIRE(area == Approx(parsedInput.cavityArea()));
```
Comment (robertodr) [inline tests/input/input_diffuse.cpp]: You can get the area from `cavityParams().area`, instead of adding a new getter to the input object.

File: tests/input/input_diffuse.cpp [hunk 10]
```diff
@@ -49,9 +49,8 @@ TEST_CASE("Input reading using GetKw for an input file for a diffuse environment
   Input parsedInput = Input(filename);
   std::string units = "ANGSTROM";
   int CODATAyear = 1998;
-  std::string type = "WAVELET";
-  int patchLevel = 1;
-  double coarsity = 0.3;
+  std::string type = "GEPOL";
+  double area = 0.3 * angstromToBohr() * angstromToBohr();
```
Comment (robertodr) [inline tests/input/input_diffuse.cpp]: There is an `angstrom2ToBohr2()` function available for this in `PhysicalConstants.cpp`

File: tools/pcmsolver.py.in [hunk 11]
```diff
@@ -217,94 +216,66 @@ def setup_keywords():
...
-    # Valid values: double in ]0.0, 1.0[
-    # Default: 0.5
-    cavity.add_kw('COARSITY',    'DBL', 0.5)
     # Average area (in au^2)
-    # Valid for: GePol and TsLess cavities
-    # Valid values: double strictly greater than 0.01 au^2
-    # Default: 0.3 au^2
+    # Valid for: GePol cavity
+    # Valid values: double strictly greater than 0.01 Angstrom^2
+    # Default: 0.3 Angstrom^2 (later converted to 1.07132... au^2)
```
Comment (robertodr) [inline tools/pcmsolver.py.in]: The minimal input assumes atomic units as defaults, as such the default for areas really is 0.3 AU^2 We should maybe re-set the default to the AU^2 value of 0.3 Angstrom^2, instead of having this oblique comment in the script.

**Comment by bast** [comment]
You can avoid the merge commit with `git pull --rebase`. Only downside of this is that if you have a conflict, then it looks a bit more hairy to resolve. But if you work on separate branches, this should not happen.

**Comment by ilfreddy** [comment]
I have followed roberto's instructions. We were just commenting about how the whole structure is getting a bit overwhelming at times (understatement)...

**Comment by bast** [comment]
What parts exactly are overwhelming? I am asking this so that we either improve docs or perhaps write up a guide or blog post.

**Comment by ilfreddy** [comment]
I will try to make a (very) long story a bit shorter.
1) Concerning this commit
I needed to rebase my branch in order to make sure that my changes would then come on top of what roberto has done with the following sequence
git pull --rebase upstream release/1.Y
git checkout claenup75
git rebase -i release/1.Y
Ignoring the fact that I would not have been able to figure this out myself, this is a bit of an overkill when I could have simply merged my branch and filed a new PR later.

2) Concerning what I am doing now
I am removing dead parts of the code which are just "littering" the released branch. Doing this correctly (right way of committing, right way of running the test, figuring out how the test work etc...) is not exactly straightforward and although I don't regard myself as very confident with all these tools (git, github, ctest, travis) I am not a complete newby either. If a task so simple requires me three full days of work, I have no chance of getting back to a state where I can confidently do meaningful code development again. As a matter of fact I also get reminded when I push something that it's good that I have done it but maybe I should have better done it in a different way, which makes me (and for sure others who are less confident than me) afraid of doing anything for the fear of doing it wrong.

3) Concerning a broader perspective
PCMSolver, MRChem or other codes should in principle be used as tools by others or accept contributions from other researchers (PhD students, postdocs). If the task of implementing anything becomes so daunting that only people who are extremely skilled professional developers can do it right (you two @bast and @robertodr are way above the average skill level I can expect to find in general) then the codes themselves are doomed the moment such people will leave the business.

**Comment by bast** [comment]
Thanks for the very good points - indeed we have to be careful that we don't push people out with our better tools and workflows.

**Comment by bast** [comment]
Just a hint for problem 1):
Assuming that you have no modifications to `release/1.Y` on your fork or local clone you could have done this (which is perhaps easier since no interactive work needed):
```
$ git fetch upstream
$ git checkout claenup75
$ git rebase upstream/release/1.Y
```

**Comment by robertodr** [comment]
For problem 1, there's a technical solution and once we agree upon it, I don't see why it will become easily second nature.
For problem 2, I guess that's my fault. Maybe the code is too contrived? Or the infrastructure around it is? I don't know. I know that I can work with it, but that doesn't matter. If it's difficult, it has to be simplified. But I can't do it on my own, because for me it's crystal clear...
For problem 3, see problem 2 above. I don't have a solution (right now, at least)

**Comment by ilfreddy** [comment]
il_freddy [11:53 AM] 
I don’t think it is your fault. I think we need to make sure that in general the required learning curve to start doing something meaningful in the code is reduced. F77 has a number of defects but one clear virtue: WYSIWYG :slightly_smiling_face: In other words, nothing hides behind the line of code you are currently reading. Knowing something more is just a print statement away.
the ctest/travis/github/git/cmake/python/c++ infrastructure is quite a lot to learn before one even attempts at writing code. And many of the tools are meant to “help” by hiding tedious details behind the curtains. Fixing a failed test requires knowledge at all such levels:
1) How is a test run and where is the output written?
2) Which executables are indeed used for the test and how do I make the test by hand
3) How is CMake producind the tests from the source code
4) Which tests are just python stuff and which ones are also executing the C++ code
5) What is hiding behind a given C++ line? (inheritance issues, include files, libraries, global stuff…)
6) How do I (properly) commit the changes?
(These are the things I stumbled upon in the last two days ;-) )

**Comment by robertodr** [comment]
Have you checked the documentation? If this stuff is missing, add it or add an issue on the repo to keep track of it.

**Comment by ilfreddy** [comment]
I seem to recall we don't have much about running test and what to do when they give wrong answers, although the documentation for writing new tests seems good to me. I will try to add something there now that my experience with these problems is fresh!

**Comment by robertodr** [comment]
I might not have kept up with the documentation work lately, that's true.

**Comment by robertodr** [comment]
@ilfreddy status on this?

**Comment by ilfreddy** [comment]
I have not done much more on it. Too busy lately.

Luca

**Comment by robertodr** [comment]
I have incorporated this changeset into #107

**Comment by robertodr** [comment]
Superseded by #107. Closing.
